src/dispatcher.rs

```diff
@@ -32,16 +33,68 @@ impl<C> CommandDispatcher<C> {
     pub fn new() -> Self {
         let mut nodes = Slab::new();
         let root = NodeKey(nodes.insert(Node::default()));
+        let commands: Vec<CommandSpec<C>> = Vec::new();
 
-        Self { nodes, root }
+        Self {
+            nodes,
+            root,
+            commands,
+        }
     }
 
     /// Registers a command to this `CommandDispatcher`.
     pub fn register(&mut self, command: impl Command<C>) -> Result<(), RegisterError>
     where
         C: 'static,
     {
-        self.append_node(self.root, command.into_root_node())
+        let spec = command.build();
+
+        let mut arguments = spec.arguments.iter();
+
+        let mut current_node_key = &self.root;
+
+        'arguments: while let Some(argument) = arguments.next() {
+            let current_node = &self.nodes[current_node_key.0];
+            for next_node_key in &current_node.next {
+                let next_node = &self.nodes[next_node_key.0];
+                match (argument, &next_node.kind) {
+                    (Argument::Literal { value }, NodeKind::Literal(node_value))
+                        if value == node_value => {
+                            current_node_key = next_node_key;
+                            continue 'arguments;
+                        }
+                    (Argument::Parser { checker, .. }, NodeKind::Parser(node_checker))
+                        if checker.equals(node_checker) => {
+                            current_node_key = next_node_key;
+                            continue 'arguments;
+                        }
+                    (_, NodeKind::Root) => panic!("?"),
+                    _ => continue,
```

Comment:
I think this should actually be `break 'arguments`, so we stop iterating once we hit a node without a matching child. Right now, this loop just consumes all the arguments, regardless of whether it hits a match or not, and as such the next loop down below never executes. So I'm not sure why the tests succeed... am I missing something?